src/github.com/getlantern/detour/detour.go

```diff
@@ -1,400 +1,162 @@
-	return atomic.LoadUint64(&dc.readBytes) > 0
-}
+		detourAllowed := eventual.NewValue()
 
-func (dc *Conn) incReadBytes(n int) {
-	atomic.AddUint64(&dc.readBytes, uint64(n))
-}
+		conn := &dconn{
+			direct: dialDirect(network, addr, isHTTP, detourAllowed),
+		}
 
-// Read() implements the function from net.Conn
-func (dc *Conn) Read(b []byte) (n int, err error) {
-	if dc.anyDataReceived() {
-		return dc.followupRead(b)
-	}
-	// At initial stage, we only have one connection,
-	// but detour connection can be available at anytime.
-	if !dc.withValidConn(func(c conn) { c.FirstRead(b, dc.chRead) }) {
-		return 0, fmt.Errorf("no connection available to %s", dc.addr)
-	}
-	for count := 1; count > 0; count-- {
-		select {
-		case newConn := <-dc.chDetourConn:
-			if atomic.LoadUint32(&dc.nonidempotentHTTPRequest) == 1 {
-				log.Tracef("Not replay nonidempotent request to %s, only add to whitelist", dc.addr)
-				AddToWl(dc.addr, false)
-				if err := newConn.Close(); err != nil {
-					log.Debugf("Error closing detour connection to %s: %s", dc.addr, err)
-				}
-				return
+		// Use direct as the initial reader
+		conn.reader = conn.direct
+
+		// Set up a detoured connection which we'll use if we have to
+		detouredTimeout := DirectDialTimeout * 2
+		conn.detoured = newEventualConn(detouredTimeout, BufferSize, func() (net.Conn, error) {
+			allowed, ok := detourAllowed.Get(detouredTimeout)
+			if !ok || !allowed.(bool) {
+				err := fmt.Errorf("Detouring not allowed for %v. Timed out waiting for connection? %v", addr, !ok)
+				log.Trace(err)
+				return nil, err
 			}
-			log.Tracef("Got detour connection to %s, replay previous op on it", dc.addr)
-			dc.muWriteBuffer.RLock()
-			sentBytes := dc.writeBuffer.Bytes()
-			dc.muWriteBuffer.RUnlock()
-			newConn.Write(sentBytes, dc.chWrite)
-			newConn.FirstRead(b, dc.chRead)
-			count++
-			// add new connection to connections
-			dc.conns <- newConn
-		case result := <-dc.chRead:
-			conn, n, err := result.conn, result.n, result.err
-			if err != nil {
-				log.Tracef("Read from %s connection to %s failed, closing: %s", typeOf(conn), dc.addr, err)
-				if err := conn.Close(); err != nil {
-					log.Debugf("Error closing %s connection to %s: %s", typeOf(conn), dc.addr, err)
-				}
-				// skip failed connection as we have more
-				if count > 1 {
-					continue
-				}
-				switch conn.ConnType() {
-				case connTypeDirect:
-					// if we haven't dial detour yet, do so now
-					select {
-					case dc.chDialDetourNow <- true:
-						count++
-					default:
-					}
-					continue
-				case connTypeDetour:
-					log.Tracef("Detour connection to %s failed, removing from whitelist", dc.addr)
-					RemoveFromWl(dc.addr)
-					// no more connections, return directly to avoid dead lock
-					return n, err
-				}
+			log.Tracef("Dialing detour for %v", addr)
+			conn, err := dialDetour(network, addr)
+			if err == nil {
+				// Prefetch up to 1 MB
+				conn = newEagerconn(conn, 1024768)
```

Comment:
@myleshorton @fffw This is an optimization that causes the detour connection to load up to 1 MB of response before we've even failed over to detour.  That way, when we decide that the direct connection is no good, we've already done a lot of the detour reading and don't need to wait for that.